### PR TITLE
[dtensor][test] test case suite for comm_mode features

### PR DESCRIPTION
**Pull request by sinhaanshul**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129017
* #128887
* __->__ #128729

**Summary**
Currently, there is only an example file for comm_mode and its features. I have created test cases that mirror the examples while the more complicated test cases also ensure that comm_mode resets all variables when used multiple times in the same function. This test case suite will also help developers ensure that new code they add to comm_mode does not affect correctness of old features. 
#128536

**Test Plan**
pytest test/distributed/_tensor/debug/test_comm_mode_features.py

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

**Review by tianyu-l — COMMENTED** [review]
please address review comments in #128536

File: test/distributed/_tensor/debug/test_comm_mode_features.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+        model2 = Transformer(model_args).to(device=self.device_type)
+        model2 = Transformer.parallelize(model2, device_mesh, is_seq_parallel)
+
+        inp_size = [8, 8]
+
+        inp = torch.randint(model_args.vocab_size, inp_size, device=self.device_type)
+
+        comm_mode = CommDebugMode()
+        with comm_mode:
+            output = model2(inp)
```
Comment (XilunWu) [inline test/distributed/_tensor/debug/test_comm_mode_features.py:257]: CI failure here.

**Review by XilunWu — COMMENTED** [review]
LGTM, please fix the CI failure

**Comment by XilunWu** [comment]
also address Tianyu's comment

File: test/distributed/_tensor/debug/test_comm_mode_features.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+    with_comms,
+)
+
+
+c10d_functional = torch.ops.c10d_functional
+
+
+class PrintModuleParamaterShardingTest(DTensorTestBase):
+    # checks if parameter / sharding info is the same as ground truth
+    def check_same_set_of_keys(self, dict1, dict2):
```
Comment (wz337) [inline test/distributed/_tensor/debug/test_comm_mode_features.py:33]: Other than the class needed to started with Test as Tianyu mentioned, I think each unit test also needed to be started with `test_`.

**Review by XilunWu — APPROVED** [review]

**Pull request closed without merging**